### PR TITLE
Feat/gpg agent forwarding

### LINKED ISSUE #580: Request: sync gpg signing keys in addition to git & docker credentials
For projects that utilize commit signing ([https://git-scm.com/book/en/v2/Git-Tools-Signing-Your-Work](https://git-scm.com/book/en/v2/Git-Tools-Signing-Your-Work)), it would be really handy to have GPG signing keys already available in the workspace.

**Is your feature request related to a problem?**

No

**Which solution do you suggest?**

Coder appears to have this option ([https://coder.com/docs/v2/latest/cli/ssh#-g---forward-gpg](https://coder.com/docs/v2/latest/cli/ssh#-g---forward-gpg)) and appears to implement it by first uploading GPG keys and then forwarding the GPG Agent: [https://github.com/coder/coder/blob/af036b440bd776f5fe57e0ddea7fc0c0596ffff7/cli/ssh.go#L297C8-L297C8](https://github.com/coder/coder/blob/af036b440bd776f5fe57e0ddea7fc0c0596ffff7/cli/ssh.go#L297C8-L297C8).

**Which alternative solutions exist?**

It looks like it may be possible to forward the GPG Agent via SSH config:
[https://unix.stackexchange.com/questions/656738/how-to-forward-gpg-agent-from-host-to-docker-container-and-reuse-cached-passphra](https://unix.stackexchange.com/questions/656738/how-to-forward-gpg-agent-from-host-to-docker-container-and-reuse-cached-passphra).

**Additional context**

### PR DESCRIPTION
**Pull request by 89luca89**
This PR Will

- migrate to our ssh server library github.com/loft-sh/ssh
- add reverse-forward-ports to ssh command
- add gpg-agent-forwarding to up command

Before merging I'd like to have feedbacks from non-linux users, as I've been able to test this only on a series of linux machines

Fixes #580 
Resolves ENG-1809

### CODE REVIEW

File: cmd/up.go [hunk 1]
```diff
@@ -228,6 +240,15 @@ func (cmd *UpCmd) Run(
 		}
 	}
 
+	// if GpgAgentForwarding we need to keep running in order to keep the reverse tunnel
+	// functioning, or gpg-agent will drop
+	if cmd.GpgAgentForwarding {
+		log.Infof(
+			"GPG Agent forwarding specified, keep this process running to have working gpg-agent forwarding",
+		)
+		select {}
```
Comment (FabianKramm) [inline cmd/up.go:249]: Why not call this at the end of up and then block?

Comment (FabianKramm) [reply cmd/up.go:249]: Would also be good to have a context variable for this so we don't need to use the flag everytime

File: cmd/up.go [hunk 2]
```diff
@@ -819,3 +840,222 @@ func setupDotfiles(
...
+echo $agent_socket
+if [ -S $agent_socket ]; then
+  echo agent socket exists, attempting to kill it >&2
+  gpgconf --kill gpg-agent
+  rm -f $agent_socket
+  sleep 1
+fi
+
+test ! -S $agent_socket
+'`)
```
Comment (FabianKramm) [inline cmd/up.go:900]: Instead of having this bash script, can we create a devpod command for this, e.g. `devpod agent workspace start-gpg-agent`

File: cmd/up.go [hunk 3]
```diff
@@ -819,3 +840,222 @@ func setupDotfiles(
...
+
+	err = exec.Command(execPath, fixAgent...).Run()
+	if err != nil {
+		return err
+	}
+
+	log.Infof("gpg-agent forwarding done")
+
+	return nil
```
Comment (FabianKramm) [inline cmd/up.go:1060]: can we condense all this into a single command that is executed? so the goal should be to only call exec.Command once

File: pkg/ssh/forward.go [hunk 4]
```diff
@@ -30,7 +36,9 @@ func PortForward(ctx context.Context, client *ssh.Client, localNetwork, localAdd
 	}()
 
 	counter := newConnectionCounter(ctx, exitAfterTimeout, func() {
-		log.Fatalf("Stopping devpod up, because it stayed idle for a while. You can disable this via 'devpod context set-options -o EXIT_AFTER_TIMEOUT=false'")
+		log.Fatalf(
+			"Stopping devpod up, because it stayed idle for a while. You can disable this via 'devpod context set-options -o EXIT_AFTER_TIMEOUT=false'",
```
Comment (FabianKramm) [inline pkg/ssh/forward.go:40]: Can you revert that?

File: pkg/ssh/forward.go [hunk 5]
```diff
@@ -86,3 +99,87 @@ func forward(localConn net.Conn, client *ssh.Client, remoteNetwork, remoteAddr s
...
+		select {
+		case <-done:
+		case <-ctx.Done():
+			_ = listener.Close()
+		}
+	}()
+
+	counter := newConnectionCounter(ctx, exitAfterTimeout, func() {
+		log.Fatalf(
+			"Stopping devpod up, because it stayed idle for a while. You can disable this via 'devpod context set-options -o EXIT_AFTER_TIMEOUT=false'",
```
Comment (FabianKramm) [inline pkg/ssh/forward.go:129]: That looks very weird, can you do a log.Fatal instead

**Review by FabianKramm — CHANGES_REQUESTED** [review]

File: pkg/ssh/forward.go (hunk 5 above)
Comment (89luca89) [reply pkg/ssh/forward.go]: Done

File: pkg/ssh/forward.go (hunk 4 above)
Comment (89luca89) [reply pkg/ssh/forward.go]: Done

File: cmd/up.go (hunk 3 above)
Comment (89luca89) [reply cmd/up.go:947]: Done

File: cmd/up.go (hunk 2 above)
Comment (89luca89) [reply cmd/up.go]: Done

**Comment by 89luca89** [comment]
Tested working on Macos also :+1:

File: docs/pages/developing-in-workspaces/credentials.mdx [hunk 6]
```diff
@@ -23,4 +23,20 @@ DevPod will make docker registry credentials available inside the dev container
 If you don't want DevPod to inject the credentials, you can disable that via the following command for all workspaces:
 ```
 devpod context set-options default -o INJECT_DOCKER_CREDENTIALS=false
-```
\ No newline at end of file
+```
+
+## GPG credentials
+
+DevPod will gpg keys available inside the dev container through an ssh tunnel. This allows you to sign commits from inside the workspace.
```
Comment (pbialon) [inline docs/pages/developing-in-workspaces/credentials.mdx:30]: "will gpg keys"? will forward?

File: pkg/ssh/forward.go [hunk 7]
```diff
@@ -86,3 +97,85 @@ func forward(localConn net.Conn, client *ssh.Client, remoteNetwork, remoteAddr s
...
+	defer listener.Close()
+
+	done := make(chan struct{})
+	defer close(done)
+
+	go func() {
+		select {
+		case <-done:
+		case <-ctx.Done():
+			_ = listener.Close()
```
Comment (pbialon) [inline pkg/ssh/forward.go:121]: this function looks exactly the same as `PortForward()` - only differs in calling `forward()` and `reverseForward()`. 

What about not duplicating the code, by reuse it?

Comment (pbialon) [reply pkg/ssh/forward.go:121]: And those two (forward/reverseForward()) also look similar. Maybe we don't need to duplicate that code as well.

File: cmd/ssh.go [hunk 8]
```diff
@@ -192,16 +226,78 @@ func (cmd *SSHCmd) jumpContainer(ctx context.Context, devPodConfig *config.Confi
...
+				ctx,
+				devPodConfig,
+				containerClient,
+				client.WorkspaceConfig().IDE.Name,
+				log,
+			)
+		})
+}
+
+func (cmd *SSHCmd) reverseForwardPorts(
```
Comment (pbialon) [inline cmd/ssh.go:245]: `reverseForwardPorts()` also looks really similar to `forwardPorts`

File: cmd/ssh.go [hunk 9]
```diff
@@ -233,10 +344,31 @@ func (cmd *SSHCmd) forwardPorts(ctx context.Context, containerClient *ssh.Client
 	return <-errChan
 }
 
-func (cmd *SSHCmd) startTunnel(ctx context.Context, devPodConfig *config.Config, containerClient *ssh.Client, ideName string, log log.Logger) error {
+func (cmd *SSHCmd) startTunnel(
```
Comment (pbialon) [inline cmd/ssh.go:347]: in general those kind of refactorings I'd put in separate PR - there is already a lot of lines of code :)

File: cmd/ssh.go [hunk 10]
```diff
@@ -284,3 +456,89 @@ func (cmd *SSHCmd) startServices(ctx context.Context, devPodConfig *config.Confi
 		}
 	}
 }
+
+// setupGPGAgent will forward a local gpg-agent into the remote container
+// this works by using cmd/agent/workspace/setup_gpg
+func (cmd *SSHCmd) setupGPGAgent(
+	ctx context.Context,
+	devPodConfig *config.Config,
```
Comment (pbialon) [inline cmd/ssh.go:464]: unused devPodConfig

File: cmd/ssh.go [hunk 11]
```diff
@@ -284,3 +456,89 @@ func (cmd *SSHCmd) startServices(ctx context.Context, devPodConfig *config.Confi
 		}
 	}
 }
+
+// setupGPGAgent will forward a local gpg-agent into the remote container
+// this works by using cmd/agent/workspace/setup_gpg
+func (cmd *SSHCmd) setupGPGAgent(
```
Comment (pbialon) [inline cmd/ssh.go:462]: if you have an idea, you can split that function - it's a bit longish :)

**Review by pbialon — DISMISSED** [review]

File: docs/pages/developing-in-workspaces/credentials.mdx (hunk 6 above)
Comment (89luca89) [reply docs/pages/developing-in-workspaces/credentials.mdx]: Good catch :)

File: pkg/ssh/forward.go (hunk 7 above)
Comment (89luca89) [reply pkg/ssh/forward.go]: Yep done now for Forward/ReverseForward

But the actual forwarding functions (forward/reverse) have quite separate concepts and connections, so I'd like to keep them separated for the sake of simplicity

File: cmd/ssh.go (hunk 8 above)
Comment (89luca89) [reply cmd/ssh.go:245]: As before yes they are similar but separate (opposite) concepts and I'd like to keep them separated

File: cmd/ssh.go (hunk 10 above)
Comment (89luca89) [reply cmd/ssh.go]: Thanks done :+1:

File: cmd/ssh.go (hunk 11 above)
Comment (89luca89) [reply cmd/ssh.go:462]: Yes it's a bit on the long side, but I don't feel like it's doing out-of-scope things, it's doing that one single thing
which happens to be a bit long :joy:

File: cmd/ssh.go [hunk 12]
```diff
@@ -248,8 +380,34 @@ func (cmd *SSHCmd) startTunnel(ctx context.Context, devPodConfig *config.Config,
...
+
+		if err == nil && len(out.Bytes()) > 1 {
+			log.Debugf("gpg: exporting already running, skipping")
+		} else {
+			err := cmd.setupGPGAgent(ctx, devPodConfig, containerClient, log)
+			if err != nil {
+				return err
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (89luca89) [reply cmd/ssh.go]: That would be a separate function called only by this function, I don't feel like it makes a lot of sense here

File: cmd/agent/workspace/setup_gpg.go [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+
+// will forward a local gpg-agent into the remote container
+// this works by
+//
+// - stopping remote gpg-agent and removing the sockets
+// - exporting local public keys and owner trust
+// - importing those into the container
+// - ensuring the gpg-agent is stopped in the container
+// - starting a reverse-tunnel of the local unix socket to remote
+// - ensuring paths and permissions are correctly set in the remote
```
[the remark this answers is not in the document]
Comment (89luca89) [reply cmd/agent/workspace/setup_gpg.go:54]: Nice idea, done :smile:

File: cmd/agent/workspace/setup_gpg.go [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+		return err
+	}
+
+	time.Sleep(time.Second)
+
+	return nil
+}
+
+func (cmd *SetupGPGCmd) stopGpgAgent() error {
+	return exec.Command("gpgconf", []string{"--kill", "gpg-agent"}...).Run()
```
Comment (pascalbreuninger) [inline cmd/agent/workspace/setup_gpg.go:122]: does this fail if the agent is already stopped?

File: cmd/ssh.go [hunk 15]
```diff
@@ -284,3 +456,88 @@ func (cmd *SSHCmd) startServices(ctx context.Context, devPodConfig *config.Confi
...
+		gpgExtraSocketPath,
+	)
+
+	command := strings.Join(forwardAgent, " ")
+
+	if cmd.User != "" && cmd.User != "root" {
+		command = fmt.Sprintf("su -c \"%s\" '%s'", command, cmd.User)
+	}
+
+	// We use start to keep this connection alive in background (hence the use of sleep infinity)
```
Comment (pascalbreuninger) [inline cmd/ssh.go:541]: This comment seems to be somewhat detached from the code, could you either elaborate or rewrite pls?

File: cmd/agent/workspace/setup_gpg.go [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+// will forward a local gpg-agent into the remote container
+// this works by
+//
+// - stopping remote gpg-agent and removing the sockets
+// - exporting local public keys and owner trust
+// - importing those into the container
+// - ensuring the gpg-agent is stopped in the container
+// - starting a reverse-tunnel of the local unix socket to remote
+// - ensuring paths and permissions are correctly set in the remote
+func (cmd *SetupGPGCmd) Run(ctx context.Context) error {
```
Comment (pascalbreuninger) [inline cmd/agent/workspace/setup_gpg.go:55]: nit: feels like there's enough in here to justify a `pkg/gpg`

File: cmd/ssh.go (hunk 8 above)
Comment (pascalbreuninger) [inline cmd/ssh.go:245]: this could potentially be invoked twice with similar arguments, is it safe to do so?

File: cmd/agent/workspace/setup_gpg.go (hunk 14 above)
Comment (89luca89) [reply cmd/agent/workspace/setup_gpg.go:122]: Nope and that's ok we want to have it stopped anyway

File: cmd/ssh.go (hunk 15 above)
Comment (89luca89) [reply cmd/ssh.go:541]: Ah it's a leftover from a previous implementation, I'll remove it

File: cmd/ssh.go (hunk 8 above)
Comment (89luca89) [reply cmd/ssh.go:245]: What do you mean? as in "run multiple times the gpg-forwarding"?

File: cmd/agent/workspace/setup_gpg.go (hunk 16 above)
Comment (89luca89) [reply cmd/agent/workspace/setup_gpg.go:51]: Done

File: cmd/ssh.go (hunk 15 above)
Comment (89luca89) [reply cmd/ssh.go]: Removed

File: cmd/ssh.go (hunk 8 above)
Comment (pascalbreuninger) [reply cmd/ssh.go:245]: nope, but if you provide the flag **and** gpg forwarding, this function will be called twice (or I missed smth)

File: pkg/ssh/server/ssh.go [hunk 17]
```diff
@@ -50,13 +51,28 @@ func NewServer(addr string, hostKey []byte, keys []ssh.PublicKey, log log.Logger
...
 			},
+			ReverseUnixForwardingCallback: func(ctx ssh.Context, socketPath string) bool {
+				log.Debugf("attempt to bind socket %s", socketPath)
+
+				_, err := os.Stat(socketPath)
+				if err == nil {
+					log.Debugf("%s already exists, removing", socketPath)
+
+					_ = os.Remove(socketPath)
+				}
```
Comment (pbialon) [inline pkg/ssh/server/ssh.go:62]: why do we delete it? And where the binding happens?

File: pkg/ssh/forward.go [hunk 18]
```diff
@@ -86,3 +141,44 @@ func forward(localConn net.Conn, client *ssh.Client, remoteNetwork, remoteAddr s
...
+		}
+	}()
+	waitGroup.Add(1)
+
+	// Copy sshConn.Reader to localConn.Writer
+	go func() {
+		defer waitGroup.Done()
+		defer remoteConn.Close()
+
+		_, err = io.Copy(remoteConn, localConn)
```
Comment (pbialon) [inline pkg/ssh/forward.go:177]: in similar cases in `loft-sh/ssh` we were using `bicopy` - why this one is different?

File: pkg/gpg/gpg_forwarding.go [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+func (g *GPGConf) SetupGpgConf() error {
+	_, err := os.Stat(filepath.Join(os.Getenv("HOME"), ".gnupg", "gpg.conf"))
+	if err != nil {
+		_, err = os.Create(filepath.Join(os.Getenv("HOME"), ".gnupg", "gpg.conf"))
+		if err != nil {
+			return err
+		}
+	}
+
+	gpgConfig, err := os.ReadFile(filepath.Join(os.Getenv("HOME"), ".gnupg", "gpg.conf"))
```
Comment (pbialon) [inline pkg/gpg/gpg_forwarding.go:62]: maybe we can have a method `configPath()` returning `filepath.Join(os.Getenv("HOME"), ".gnupg", "gpg.conf")`

File: pkg/gpg/gpg_forwarding.go [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+
+		if _, err := f.WriteString("use-agent\n"); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (g *GPGConf) SetupRemoteSocketDirtree() error {
```
Comment (pbialon) [inline pkg/gpg/gpg_forwarding.go:83]:
```suggestion
func (g *GPGConf) SetupRemoteSocketDirTree() error {
```

File: pkg/gpg/gpg_forwarding.go [hunk 21]
```diff
@@ -0,0 +1,126 @@
...
+		"chown",
+		"-R",
+		strconv.Itoa(os.Getuid())+":"+strconv.Itoa(os.Getgid()),
+		"/run/user",
+		filepath.Dir(g.SocketPath),
+		g.SocketPath,
+	).Run()
+}
+
+func (g *GPGConf) SetupRemoteSocketLink() error {
```
Comment (pbialon) [inline pkg/gpg/gpg_forwarding.go:99]: maybe a better structure or comment what that function is doing? It's not clear at the first glance, what is happening to those symlinks etc :)

File: cmd/agent/workspace/setup_gpg.go [hunk 22]
```diff
@@ -0,0 +1,121 @@
...
+		return err
+	}
+
+	logger.Debugf("Setup gpg.conf")
+	err = gpgConf.SetupGpgConf()
+	if err != nil {
+		return err
+	}
+
+	time.Sleep(time.Second)
```
Comment (pbialon) [inline cmd/agent/workspace/setup_gpg.go:118]: why do we need that sleep?

File: cmd/ssh.go [hunk 23]
```diff
@@ -192,16 +226,78 @@ func (cmd *SSHCmd) jumpContainer(ctx context.Context, devPodConfig *config.Confi
...
+		timeout, err = time.ParseDuration(cmd.ForwardPortsTimeout)
+		if err != nil {
+			return fmt.Errorf("parse forward ports timeout: %w", err)
+		}
 
-		// start ssh tunnel
-		return cmd.startTunnel(ctx, devPodConfig, containerClient, client.WorkspaceConfig().IDE.Name, log)
-	})
+		log.Infof("Using port forwarding timeout of %s", cmd.ForwardPortsTimeout)
+	}
```
Comment (pbialon) [inline cmd/ssh.go:259]: move parsing of timeout out to separate function

File: pkg/ssh/forward.go [hunk 24]
```diff
@@ -11,13 +11,63 @@ import (
...
+		remoteNetwork, remoteAddr, localNetwork, localAddr,
+		exitAfterTimeout, log, reverseForward,
+	)
+}
+
+func portForwarding(
+	ctx context.Context,
+	client *ssh.Client,
+	listener net.Listener,
+	srcNetwork, srcAddr, dstNetwork, dstAddr string,
```
Comment (pbialon) [inline pkg/ssh/forward.go:66]: `srcNetwork` is not used

File: cmd/ssh.go [hunk 25]
```diff
@@ -192,16 +226,78 @@ func (cmd *SSHCmd) jumpContainer(ctx context.Context, devPodConfig *config.Confi
...
+
+	errChan := make(chan error, len(cmd.ReverseForwardPorts))
```
Comment (pbialon) [inline cmd/ssh.go:261]: question: why do we want to have more than one reverse forward port?

Is it because of multiple workspaces using that feature or sth which I don't see?

File: cmd/ssh.go [hunk 26]
```diff
@@ -284,3 +456,87 @@ func (cmd *SSHCmd) startServices(ctx context.Context, devPodConfig *config.Confi
...
+	log log.Logger,
+) error {
+	writer := log.ErrorStreamOnly().Writer(logrus.InfoLevel, false)
+	defer writer.Close()
+
+	log.Debugf("gpg: exporting gpg public key from host")
+
+	// Read the user's public keys and ownertrust from GPG.
+	// These commands are executed LOCALLY, the output will be imported by the remote gpg
+	pubKeyExport, err := exec.Command("gpg", "--armor", "--export").Output()
```
Comment (pbialon) [inline cmd/ssh.go:474]: why won't we move those commands (this, and ones below) to gpg package? 

I feel more comfortable with a "client" calling `exec.Command` underneath instead of calling in directly in hear.

File: cmd/ssh.go [hunk 27]
```diff
@@ -248,8 +380,34 @@ func (cmd *SSHCmd) startTunnel(ctx context.Context, devPodConfig *config.Config,
...
+		// Check if a forwarding is already enabled and running, in that case
+		// we skip the forwarding and keep using the original one
+		command := "gpg -K"
+		if cmd.User != "" && cmd.User != "root" {
+			command = fmt.Sprintf("su -c \"%s\" '%s'", command, cmd.User)
+		}
+
+		// capture the output, if it's empty it means we don't have gpg-forwarding
+		var out bytes.Buffer
+		err := devssh.Run(ctx, containerClient, command, nil, &out, writer)
```
Comment (pbialon) [inline cmd/ssh.go:394]: maybe encapsulate it somehow, with `remoteGpgClient`? For me it's too low-level abstraction we use for the function which is responsible for checking if gpg forwarding is already set up and if not then to do it.

File: cmd/ssh.go (hunk 8 above)
Comment (89luca89) [reply cmd/ssh.go:245]: Tested and it indeed created problems (sometimes)
Fixed it now thanks!

File: pkg/ssh/server/ssh.go (hunk 17 above)
Comment (89luca89) [reply pkg/ssh/server/ssh.go:62]: That's the behaviour of OpenSSH with the `StreamLocalBindUnlink=yes`  option
Which is useful in our case to ensure we don't have dangling (old) sockets

File: pkg/ssh/forward.go (hunk 18 above)
Comment (89luca89) [reply pkg/ssh/forward.go:177]: Not sure here, might want to ask @FabianKramm (the reverse code is a copy of the regular forward one more or less)

File: pkg/gpg/gpg_forwarding.go (hunk 19 above)
Comment (89luca89) [reply pkg/gpg/gpg_forwarding.go]: Done :+1:

File: cmd/agent/workspace/setup_gpg.go (hunk 22 above)
Comment (89luca89) [reply cmd/agent/workspace/setup_gpg.go]: Debugging leftover, removed :+1:

File: pkg/gpg/gpg_forwarding.go (hunk 21 above)
Comment (89luca89) [reply pkg/gpg/gpg_forwarding.go:104]: Done :+1:

File: pkg/gpg/gpg_forwarding.go (hunk 20 above)
Comment (89luca89) [reply pkg/gpg/gpg_forwarding.go]: Done :+1:

File: pkg/ssh/forward.go (hunk 24 above)
Comment (89luca89) [reply pkg/ssh/forward.go:66]: That's needed in order for `ForwardingFunction` alias/type to work and return the function (eithe regular or reverse forward) to be returned to `portForwarding`

File: cmd/ssh.go (hunk 25 above)
Comment (89luca89) [reply cmd/ssh.go:261]: Yes, it may be that one wants to forward more ports, sockets or whatnot at once

File: cmd/ssh.go (hunk 23 above)
Comment (89luca89) [reply cmd/ssh.go]: Done :+1:

File: cmd/ssh.go (hunk 26 above)
Comment (89luca89) [reply cmd/ssh.go]: Done :+1:

File: cmd/ssh.go (hunk 27 above)
Comment (89luca89) [reply cmd/ssh.go]: Moved to the gpg package :+1:

**Review by FabianKramm — CHANGES_REQUESTED** [review]
@89luca89 can we add a unit or e2e test for this? Since this is something thats default disabled and we have changes on the roadmap for the underlying ssh connection, we need to make sure those future changes are not breaking this

**Comment by 89luca89** [comment]
> @89luca89 can we add a unit or e2e test for this? Since this is something thats default disabled and we have changes on the roadmap for the underlying ssh connection, we need to make sure those future changes are not breaking this

:+1: sure thing

**Comment by FabianKramm** [comment]
Besides that it looks good, but we need a rebase here

**Review by FabianKramm — APPROVED** [review]
